include/uart.h

```diff
@@ -29,6 +29,157 @@ extern "C" {
+ *   |      |                            need another buffer)
+ *   |      |
+ *   |      UART_RX_RDY (Timeout occurred, received n bytes)
+ *   |
+ *   UART_RX_BUF_REQUEST (Started transmission to first buffer,
+ *                        need second buffer)
+ *
+ */
+enum uart_event_type {
+	/** @brief Whole TX buffer was transmitted. */
+	UART_TX_DONE,
+	/** @brief TX timed out.
+	 *
+	 * When flow control is enabled, there is a possibility that TX transfer
+	 * won't finish.
+	 */
+	UART_TX_TIMEOUT,
+	/**
+	 * @brief Received data is ready for processing.
+	 *
+	 * This event is generated in two cases:
+	 * - When RX timeout occurred, and data was stored in provided buffer.
+	 * - When provided buffer is full.
+	 */
+	UART_RX_RDY,
+	/**
+	 * @brief Driver requests next buffer for continuous transmission.
+	 *
+	 * This event is triggered when receiving has started for a new buffer.
+	 * For continuous reliable receiving, user should provide another RX
+	 * buffer in response to this event, using uart_rx_buf_rsp() function.
+	 *
+	 * If uart_rx_buf_rsp() function is not called, receiving will stop
+	 * after filling current buffer and UART_RX_DONE will be generated.
+	 */
+	UART_RX_BUF_REQUEST,
+	/**
+	 * @brief Receiving has stopped and provided buffer is full.
+	 *
+	 * If no second buffer was provided, receiving will stop after filling
+	 * current buffer. After that event, receiving can be enabled again
+	 * using rx_en() function.
+	 */
+	UART_RX_DONE,
+	/**
+	 * @brief Buffer is no longer used by UART driver.
+	 *
+	 * This event occurs after rx_disable() function is called, for every
+	 * buffer passed to UART driver, it indicates that buffer can be freed
+	 * and is no longer used.
+	 */
+	UART_RX_BUF_RELEASED,
+	/**
+	 * @brief RX has been disabled and can be reenabled.
+	 *
+	 * After rx_disable() function is called, driver may need some time to
+	 * stop the transmission, after releasing buffers this event is
+	 * generated and receiving can be enabled again using rx_en() function.
+	 */
+	UART_RX_DISABLED,
+	/**
+	 * @brief Break interrupt
+	 *
+	 * A break interrupt was received. This happens when the serial input
+	 * is held at a logic '0' state for longer than the sum of
+	 * start time + data bits + parity + stop bits.
+	 */
+	UART_BREAK,
+	/** @brief Overrun error */
+	UART_ERROR_OVERRUN,
+	/** @brief Parity error */
+	UART_ERROR_PARITY,
+	/** @brief Framing error */
+	UART_ERROR_FRAMING,
+};
+
+/* @brief Backward compatibility defines, deprecated */
+#define UART_ERROR_BREAK UART_BREAK
+#define LINE_CTRL_BAUD_RATE (1 << 0)
+#define LINE_CTRL_RTS UART_LINE_CTRL_RTS
+#define LINE_CTRL_DTR UART_LINE_CTRL_DTR
+#define LINE_CTRL_DCD UART_LINE_CTRL_DCD
+#define LINE_CTRL_DSR UART_LINE_CTRL_DSR
+
+/* @brief Structure containing information about current event. */
+struct uart_event {
+	/* @brief Type of event */
+	enum uart_event_type type;
+	/* @brief Pointer to current buffer. */
+	u8_t *buf;
```

Comment:
as different events have different event content, i would turn it into union since now it's not clear if for example TX DONE has valid `*buf` or not.